module/apms3session/session.go

```diff
@@ -0,0 +1,153 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apms3session // import "go.elastic.co/apm/module/apms3session"
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+	"go.elastic.co/apm/stacktrace"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+func init() {
+	stacktrace.RegisterLibraryPackage(
+		"github.com/aws/aws-sdk-go/aws/request",
+		"github.com/aws/aws-sdk-go/aws/session",
+	)
+}
+
+type handlers struct {
+	tracer *apm.Tracer
+}
+
+// Option sets options for tracing s3 requests.
+type Option func(*handlers)
+
+// WithTracer returns an Option which sets t as the tracer to use for tracing
+// s3 requests.
+func WithTracer(t *apm.Tracer) Option {
+	if t == nil {
+		panic("t == nil")
+	}
+	return func(h *handlers) {
+		h.tracer = t
+	}
+}
+
+// WrapSession wraps the provided s3 session with handlers that hook into the
+// aws sdk's request lifecycle.
+func WrapSession(s *session.Session, opts ...Option) *session.Session {
+	h := &handlers{
+		tracer: apm.DefaultTracer,
+	}
+	for _, o := range opts {
+		o(h)
+	}
+
+	s.Handlers.Send.PushFrontNamed(request.NamedHandler{
+		Name: "go.elastic.co/apm/module/apms3session/send",
+		Fn:   h.send,
+	})
+	s.Handlers.Complete.PushBackNamed(request.NamedHandler{
+		Name: "go.elastic.co/apm/module/apms3session/complete",
+		Fn:   h.complete,
+	})
+
+	return s
+}
+
+const (
+	spanType    = "storage"
+	spanSubtype = "s3"
+)
+
+func (h *handlers) send(req *request.Request) {
+	if req.RetryCount > 0 {
+		return
+	}
+
```

Comment:
For now can we check `req.ClientInfo.ServiceName == "s3"` and bail out if it's not? Let me know if you disagree: I think it would make sense to have a single module that instruments all (or at least multiple) AWS services. Later on we can switch on the service name to determine the appropriate span type, subtype, name, etc.